addons/im_livechat/static/src/legacy/public_livechat_chatbot.js

```diff
@@ -0,0 +1,882 @@
+                    nextStepDelay = 0;
+                } else {
+                    this._chatbotSetIsTyping();
+                }
+
+                this.nextStepTimeout = setTimeout(
+                    this._chatbotTriggerNextStep.bind(this), nextStepDelay);
+            }
+        }
+
+        if (!this._chatbotIsDisplayRestartButton()) {
+            this._chatWindow.$('.o_livechat_chatbot_main_restart').addClass('d-none');
+        }
+     },
+     /**
+      * See '_chatbotSaveSession'.
+      *
+      * We retrieve the livechat uuid from the session cookie since the livechat Widget is not yet
+      * initialized when we restore the chatbot state.
+      *
+      * We also clear any older keys that store a previously saved chatbot session.
+      * (In that case we clear the actual browser's local storage, we don't use the localStorage
+      * object as it does not allow browsing existing keys, see 'local_storage.js'.)
+      *
+      * @private
+      */
+    _chatbotRestoreSession: function (sessionCookie) {
+        const sessionKey = 'im_livechat.chatbot.state.uuid_' + JSON.parse(sessionCookie).uuid;
+        const browserLocalStorage = window.localStorage;
+        if (browserLocalStorage && browserLocalStorage.length) {
+            for (let i = 0; i < browserLocalStorage.length; i++) {
+                const key = browserLocalStorage.key(i);
+                if (key.startsWith('im_livechat.chatbot.state.uuid_') && key !== sessionKey) {
+                    browserLocalStorage.removeItem(key);
+                }
+            }
+        }
+
+        let chatbotState = localStorage.getItem(sessionKey);
+
+        if (chatbotState) {
+            chatbotState = JSON.parse(chatbotState);
+            this._isChatbot = true;
+            this._chatbot = chatbotState._chatbot;
+            this._chatbotCurrentStep = chatbotState._chatbotCurrentStep;
+        }
+     },
+    /**
+     * Register current chatbot step state into localStorage to be able to resume if the visitor
+     * goes to another website page or if he refreshes his page.
+     *
+     * (Will not work if the visitor switches browser but his livechat session will not be restored
+     *  anyway in that case, since it's stored into a cookie).
+     *
+     * @private
+     */
+    _chatbotSaveSession: function () {
+        const chatUuid = this._livechat.toData().uuid;
+        localStorage.setItem('im_livechat.chatbot.state.uuid_' + chatUuid, JSON.stringify({
+            '_chatbot': this._chatbot,
+            '_chatbotCurrentStep': this._chatbotCurrentStep,
+        }));
+    },
+    /**
+     * Adds a small "is typing" animation into the chat window.
+     *
+     * @private
+     */
+    _chatbotSetIsTyping: function (isWelcomeMessage=false) {
+        if (this.isTypingTimeout) {
+            clearTimeout(this.isTypingTimeout);
+        }
+
+        this._chatbotDisableInput('');
+
+        const avatarSource = this._chatbot.chatbot_has_avatar
```

Comment:
Can't we have an avatar_url computed beforehand to avoid handcrafting it here ? Either in model, controller, or at init of chatbot ?